src/main/java/net/neoforged/neoforge/transfer/handlers/templates/resources/ItemAccessResourceHandler.java

```diff
@@ -0,0 +1,170 @@
+
+    protected ItemAccessResourceHandler(ItemAccess itemAccess, int size) {
+        this.itemAccess = itemAccess;
+        this.size = size;
+    }
+
+    /**
+     * Retrieves the resource stored in the {@linkplain ItemAccess#getResource() current contents} of the item access.
+     */
+    protected abstract T getResourceFrom(ItemResource accessResource, int index);
+
+    /**
+     * Retrieves the amount stored in the {@linkplain ItemAccess#getResource() current contents} of the item access.
+     */
+    protected abstract int getAmountFrom(ItemResource accessResource, int index);
+
+    /**
+     * Returns a resource with updated resource and amount.
+     *
+     * @param accessResource current resource, before the update
+     * @param index          the index at which the resource and amount should be updated
+     * @param newResource    the new resource, <strong>never empty</strong>: empty is indicated by a 0 amount
+     * @param newAmount      the new amount
+     * @return {@code accessResource} updated with the new resource and amount,
+     *         or {@link ItemResource#EMPTY} if the new resource or amount cannot be stored
+     * @implNote This function <strong>should not</strong> mutate the {@linkplain #itemAccess item access},
+     *           that will be done by the calling code based on the results of this function.
+     */
+    // TODO: we could allow returning null when the resource/amount should be deleted
+    // TODO: this would allow for "consumable" implementations with minimal effort
+    protected abstract ItemResource update(ItemResource accessResource, int index, T newResource, int newAmount);
+
+    /**
+     * Return {@code true} if the passed non-empty resource can fit in this handler, {@code false} otherwise.
+     *
+     * <p>The result of this function is used in the provided implementations of:
+     * <ul>
+     * <li>{@link #getCapacityAsLong(int, T)}, to report a capacity of {@code 0} for invalid items;</li>
+     * <li>{@link #insert(int, T, int, TransactionContext)}, to reject items that cannot fit in this handler.</li>
+     * </ul>
+     */
+    @Override
+    public boolean isValid(int index, T resource) {
+        return true;
+    }
+
+    /**
+     * Return the maximum capacity of this handler for the passed resource.
+     * If the passed resource is empty, an estimate should be returned.
+     *
+     * @return The maximum capacity of this handler for the passed resource.
+     */
+    protected abstract int getCapacity(int index, T resource);
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public T getResource(int index) {
+        Objects.checkIndex(index, size());
+        return getResourceFrom(itemAccess.getResource(), index);
+    }
+
+    @Override
+    public long getAmountAsLong(int index) {
+        Objects.checkIndex(index, size());
+        return (long) itemAccess.getAmount() * getAmountFrom(itemAccess.getResource(), index);
```

Comment:
If we are allowing "oversized" stacks can we add a comment that we can just cast as MAX_INT * MAX_INT < MAX_LONG so there is no worry that it will overflow?